image-toolkit/sbin/startup.sh

```diff
@@ -48,24 +48,59 @@ vlc_restart_loop() {
                 dst=std{access=http,mux=ts,dst=:101} \
             }"
         echo "VLC exited, restarting in 3 seconds"
+
         # Sleep to prevent CPU hogging and let the processes be killed in any order
         sleep 3
     done
 }
 
 webcam_stream_loop() {
-    VIDEO_DEVICE_NO=0
-    # We might need to use a different device
-    # Find out the correct device using aplay -L
-    AUDIO_DEVICE=alsa://plughw:0,0
-    while :
-    do
-        if ! [[ -e "/dev/video$VIDEO_DEVICE_NO" ]]; then
-            continue
-        fi
+    webcam_pick_devices() {
+        # Find device by unique identifiers
+        # https://docs.kernel.org/userspace-api/media/v4l/open.html#v4l2-device-node-naming
+        while :
+        do
+            echo "Looking for video devices"
+
+            local VIDEO_DEVICES
+            mapfile -t VIDEO_DEVICES < <(find /dev/v4l/by-id -regex ".*/usb-.*-video-index0")
+
+            local VIDEO_DEVICES_COUNT
+            VIDEO_DEVICES_COUNT=${#VIDEO_DEVICES[@]}
+
+            # Zero video devices
+            if [[ $VIDEO_DEVICES_COUNT -eq 0 ]]; then
+                echo "No video devices found"
+                sleep 3
+                continue
+            else
+                echo "Found $VIDEO_DEVICES_COUNT device(s): ${VIDEO_DEVICES[@]}"
+            fi
+
+            VIDEO_DEVICE_PATH=${VIDEO_DEVICES[0]}
+            echo "Using $VIDEO_DEVICE_PATH"
+
+            # We might need to use a different device
+            # Find out the correct device using aplay -L
+            AUDIO_DEVICE=alsa://plughw:0,0
+            return
+        done
+    }
+
+    webcam_stream() {
+        # Monitor the video device using udevadm monitor
+        # If device is unplugged, kill existing clvc instance to release /dev/video0
+        echo "Starting udevadm to monitor video device connection"
+        udevadm monitor --udev -s video4linux $VIDEO_DEVICE_PATH | while read -r line; do
+            if [[ "$line" =~ "remove" ]] ; then
+                echo "Received video device removal: $line"
+                break
+            fi
+        done &
```

Comment:
Flag -s (--subsystem-match) is only used to match for subsystem, so passing $VIDEO_DEVICE_PATH will have no effect.

The pipeline will only terminate when all commands in the pipeline terminate. When we encounter a remove event, the latter process exits, but udevadm monitor continues to run until it tries to write to stdout and fails.Subsequently, vlc is not killed until the reconnect event occurs. We can avoid this behaviour using process substitution: `while read -r line; do ... done < <(udevadm monitor --udev -s video4linux)`.

Reply:
Thx. Fixed